lib/snitch-core/src/lib.rs

```diff
@@ -0,0 +1,440 @@
+        core::iter::from_fn(move || {
+            let len = u16::from_le_bytes(take_array(&mut slices)?) as usize;
+            let tid = u16::from_le_bytes(take_array(&mut slices)?);
+            let timestamp = u64::from_le_bytes(take_array(&mut slices)?);
+            let slices = take_slice(&mut slices, len)?;
+            let ena = next_ena;
+            next_ena += 1;
+            Some(Record {
+                ena,
+                tid,
+                timestamp,
+                slices,
+            })
+        })
+    }
+
+    /// Reads records starting from `first_ena`, inclusive.
+    pub fn read_from(
+        &mut self,
+        first_ena: u64,
+    ) -> impl Iterator<Item = Record<'_>> + '_ {
+        // Attempt to recover and insert a loss record if necessary.
+        self.recover_if_required();
+
+        // We can use `skip_while` instead of `filter` here because the ENAs are
+        // always ascending.
+        self.iter_contents()
+            .skip_while(move |rec| rec.ena < first_ena)
+    }
+
+    /// Discards records up through and including `last_written_ena`.
+    ///
+    /// This is intended to be called when we've received confirmation that a
+    /// range of records has been committed to a database. The provided ENA is
+    /// _inclusive_ and indicates the last record written to the database.
+    ///
+    /// If the ENA is either lower than any of our stored records, or higher
+    /// than what we've vended out, this is a no-op.
+    pub fn flush_thru(&mut self, last_written_ena: u64) {
+        let count = last_written_ena
+            .saturating_add(1)
+            .saturating_sub(self.earliest_ena);
+        for _ in 0..count {
+            // Discard the lead record in the queue.
+            let mut slices = self.storage.as_slices();
+            let len = u16::from_le_bytes(take_array(&mut slices).unwrap_lite())
+                as usize;
+
+            let size = OVERHEAD + len;
+            // So it's weird, but, heapless::Deque only lets you pop individual
+            // bytes. Hopefully this is not too expensive.
+            for _ in 0..size {
+                self.storage.pop_front();
+            }
+
+            self.earliest_ena += 1;
+        }
+    }
+
+    /// Makes a best effort at inserting a record.
+    ///
+    /// This will attempt to recover from the `Losing` state before writing
+    /// anything.
+    ///
+    /// If, after recovery, we are `Collecting` and there is enough space for a
+    /// header (`OVERHEAD`) plus `data`, the record is stored, implicitly
+    /// assigned the next ENA.
+    ///
+    /// If recovery is not possible, the existing loss count is advanced.
+    ///
+    /// If we are `Collecting` but `data` plus a header won't fit, we enter the
+    /// `Losing` state with a count of 1.
+    fn insert_impl(&mut self, sender: u16, timestamp: u64, data: &[u8]) {
+        let data_len = u16::try_from(data.len()).unwrap_lite();
```

Comment:
This is a caller-controlled panic; should we be inserting a loss record instead?

Reply:
Hm. So my initial thinking was that this is an explicit precondition that could be checked by the caller (packrat, currently), so a panic made sense. But I think you have a point. I'll try changing it and see how the client code feels.

Reply:
Alright, I've switched this logic to automatically "lose" any such packets, not that they will exist in practice.